### PR TITLE
Media: Adjust `needsProxy` check for poster images for base color generation

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

For hotlinked videos, poster images are uploaded to the media library. Meaning that poster could be served from mysite.com and video from an external site. Meaning that needsproxy on the resource only covers the asset not the poster.  

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [ ] I have added a matching `Type: XYZ` label to the PR

---

Fixes #

### CODE REVIEW

File: packages/story-editor/src/utils/useCORSProxy.js [hunk 1]
```diff
@@ -30,10 +30,38 @@ function useCORSProxy() {
     actions: { getProxyUrl },
   } = useAPI();
   const enableCORSProxy = useFeature('enableCORSProxy');
+
```
Comment (spacedmonkey) [inline packages/story-editor/src/utils/useCORSProxy.js:33]: This change is wholesale copied from https://github.com/google/web-stories-wp/pull/9713.

Comment (swissspidy) [reply packages/story-editor/src/utils/useCORSProxy.js:33]: If you add this here you also need to remove it from `useInsert`

File: packages/story-editor/src/utils/useCORSProxy.js [hunk 2]
```diff
@@ -30,10 +30,38 @@ function useCORSProxy() {
...
+    }
+
+    return shouldProxy;
+  }
+
   const getProxiedUrl = useCallback(
     (resource, src) => {
       const { needsProxy } = resource;
-      if (enableCORSProxy && needsProxy) {
+      if (enableCORSProxy && needsProxy && src) {
```
Comment (swissspidy) [inline packages/story-editor/src/utils/useCORSProxy.js:64]: I don't particularly like this. If `src` is falsey/null, `getProxiedUrl` should probably not be called on it in the first place

File: packages/story-editor/src/components/canvas/useInsertElement.js [hunk 3]
```diff
@@ -71,7 +71,8 @@ function useInsertElement() {
       }
 
       try {
-        const imageSrcProxied = getProxiedUrl(resource, imageSrc);
+        const needsProxy = await checkResourceAccess(imageSrc);
```
Comment (swissspidy) [inline packages/story-editor/src/components/canvas/useInsertElement.js]: Poster images are never hotlinked, so they don't need that. We can really avoid that HTTP request.

File: packages/story-editor/src/components/canvas/useInsertElement.js [hunk 4]
```diff
@@ -47,7 +47,7 @@ function useInsertElement() {
   const {
     actions: { registerUsage },
   } = useMedia3pApi();
-  const { getProxiedUrl } = useCORSProxy();
+  const { getProxiedUrl, checkResourceAccess } = useCORSProxy();
```
Comment (swissspidy) [inline packages/story-editor/src/components/canvas/useInsertElement.js:50]: Note that lint error in this file now

File: packages/story-editor/src/components/canvas/useInsertElement.js (hunk 3 above)
Comment (swissspidy) [reply packages/story-editor/src/components/canvas/useInsertElement.js]:
```suggestion
        const needsProxy = type === 'image' && await checkResourceAccess(imageSrc);
```

a simple fix, albeit not  the best

Comment (spacedmonkey) [reply packages/story-editor/src/components/canvas/useInsertElement.js]: No this is the wrong way around. We want to check for poster images here not images. A poster image on an hot linked image could is a local image ( not needing a proxy ) but a poster image on a coverr or tenor resource may need a proxy.

File: packages/story-editor/src/utils/useCORSProxy.js (hunk 2 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/useCORSProxy.js:64]: I know, but there is nothing wrong with a little bit of harden here. This came up with a issue in unit tests.

File: packages/story-editor/src/components/canvas/useInsertElement.js (hunk 3 above)
Comment (swissspidy) [reply packages/story-editor/src/components/canvas/useInsertElement.js]: 3P media never needs a proxy

File: packages/story-editor/src/utils/useCORSProxy.js (hunk 2 above)
Comment (swissspidy) [reply packages/story-editor/src/utils/useCORSProxy.js:64]: We should update the tests to use proper data

File: packages/story-editor/src/components/canvas/useInsertElement.js (hunk 3 above)
Comment (spacedmonkey) [reply packages/story-editor/src/components/canvas/useInsertElement.js]: Okay, but we dont have a simple check for is hotlinked.

Comment (swissspidy) [reply packages/story-editor/src/components/canvas/useInsertElement.js]: Sounds like adding `isHotlinked` to resources will make our lives a lot easier

Comment (spacedmonkey) [reply packages/story-editor/src/components/canvas/useInsertElement.js]: > Sounds like adding isHotlinked to resources will make our lives a lot easier 

Let's no add another property like this. We already `local` and `isExternal` which are boolean. We should just add `source`. Values would be `local`, `library`, `hotlink`, `unsplash`, `tenor` or `coverr`. We could then even get rid of  `local` and `isExternal`. We can even add a little helper function called, isExternalResource

File: packages/story-editor/src/utils/useCORSProxy.js (hunk 2 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/useCORSProxy.js:64]: So this function should only work if propper data is passed to it? We should not handle invalid data?

File: packages/story-editor/src/components/canvas/useInsertElement.js (hunk 4 above)
Comment (spacedmonkey) [reply packages/story-editor/src/components/canvas/useInsertElement.js:50]: ?

Comment (swissspidy) [reply packages/story-editor/src/components/canvas/useInsertElement.js:50]: There was a lint error in this file

https://github.com/google/web-stories-wp/runs/4266687792?check_suite_focus=true

You just fixed it in 76425572f524e1715ba3b89b9bfe40f85e767008

File: packages/story-editor/src/components/canvas/useInsertElement.js (hunk 3 above)
Comment (swissspidy) [reply packages/story-editor/src/components/canvas/useInsertElement.js]: Interesting suggestion. Something I'd be willing to explore in a dedicated ticket.

File: packages/story-editor/src/utils/useCORSProxy.js (hunk 2 above)
Comment (swissspidy) [reply packages/story-editor/src/utils/useCORSProxy.js:64]: That should be handled at a different level

Comment (swissspidy) [reply packages/story-editor/src/utils/useCORSProxy.js:64]: If you could point to the unit test you mentioned, that would be helpful.

File: packages/story-editor/src/components/canvas/useInsertElement.js [hunk 5]
```diff
@@ -71,7 +71,11 @@ function useInsertElement() {
       }
 
       try {
-        const imageSrcProxied = getProxiedUrl(resource, imageSrc);
+        const needsProxy =
+          type === 'image'
+            ? resource?.needsProxy
+            : await checkResourceAccess(imageSrc);
+        const imageSrcProxied = getProxiedUrl({ needsProxy }, imageSrc);
```
Comment (swissspidy) [inline packages/story-editor/src/components/canvas/useInsertElement.js:78]: Why not just

```js
const imageSrcProxied =
          type === 'video' ? imageSrc : getProxiedUrl(resource, imageSrc);
```

Comment (swissspidy) [reply packages/story-editor/src/components/canvas/useInsertElement.js:78]: I don't see the need for `checkResourceAccess` when we already have `resource.needsProxy`

Comment (spacedmonkey) [reply packages/story-editor/src/components/canvas/useInsertElement.js:78]: What if you are filtering in ( like we know some third parties are doing ) from a different media source? The fact is, poster and video could be served from different domains. It is wise here to be careful and just call `checkResourceAccess` for posters, as we do not know if they need a proxy or not. 

> const imageSrcProxied = type === 'video' ? imageSrc : getProxiedUrl(resource, imageSrc);

This feels extremely brittle.

File: packages/story-editor/src/utils/useCORSProxy.js (hunk 2 above)
Comment (spacedmonkey) [reply packages/story-editor/src/utils/useCORSProxy.js:64]: > That should be handled at a different level

That isn't DRY at all? And may result in someone forget to add this check upstream.

**Pull request merged**